File: backend/maava-new/Backend/src/modules/food/admin/models/offer.model.js
import mongoose from 'mongoose';
import { verticalPlugin } from '../../../../core/vertical/verticalScope.js';

const foodOfferSchema = new mongoose.Schema(
    {
        couponCode: { type: String, required: true, trim: true, uppercase: true },
        discountType: { type: String, enum: ['percentage', 'flat-price'], default: 'percentage', index: true },
        discountValue: { type: Number, required: true, min: 0 },
        customerScope: { type: String, enum: ['all', 'first-time'], default: 'all', index: true },
        restaurantScope: { type: String, enum: ['all', 'selected'], default: 'all', index: true },
        restaurantId: { type: mongoose.Schema.Types.ObjectId, ref: 'FoodRestaurant' },
        restaurantIds: [{ type: mongoose.Schema.Types.ObjectId, ref: 'FoodRestaurant' }],
        minOrderValue: { type: Number, default: 0, min: 0 },
        maxDiscount: { type: Number, default: null, min: 0 },
        usageLimit: { type: Number, default: null, min: 0 },
        perUserLimit: { type: Number, default: null, min: 0 },
        usedCount: { type: Number, default: 0, min: 0 },
        startDate: { type: Date },
        isFirstOrderOnly: { type: Boolean, default: false },
        endDate: { type: Date },
        status: { type: String, enum: ['active', 'paused', 'inactive'], default: 'active', index: true },
        showInCart: { type: Boolean, default: true },
        createdByRole: { type: String, enum: ['ADMIN', 'RESTAURANT'], default: 'ADMIN', index: true },
        adminBearPercentage: { type: Number, default: 100, min: 0, max: 100 },
        restaurantBearPercentage: { type: Number, default: 0, min: 0, max: 100 }
    },
    { collection: 'food_offers', timestamps: true }
);

foodOfferSchema.plugin(verticalPlugin);

/**
 * A coupon code is unique per vertical, not globally: SAVE50 on groceries and
 * SAVE50 on restaurant food are two different offers with two different budgets.
 *
 * ponytail: the legacy `couponCode_1` index still exists on deployed clusters;
 * scripts/merge-databases.js drops it.
 */
foodOfferSchema.index({ vertical: 1, couponCode: 1 }, { unique: true });

foodOfferSchema.index({ vertical: 1, restaurantId: 1, createdAt: -1 });
foodOfferSchema.index({ vertical: 1, restaurantIds: 1, createdAt: -1 });

export const FoodOffer = mongoose.model('FoodOffer', foodOfferSchema);

File: maava_restaurant/lib/features/offers/domain/offer_model.dart
class OfferModel {
  OfferModel({
    required this.id,
    required this.couponCode,
    required this.discountType,
    required this.discountValue,
    required this.minOrderValue,
    required this.maxDiscount,
    required this.usageLimit,
    required this.perUserLimit,
    required this.isFirstOrderOnly,
    required this.startDate,
    required this.endDate,
    required this.status,
  });

  factory OfferModel.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic v) =>
        v == null ? null : DateTime.tryParse(v.toString());
    num? asNum(dynamic v) => v is num ? v : num.tryParse((v ?? '').toString());

    return OfferModel(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      couponCode: (json['couponCode'] ?? '').toString(),
      discountType: (json['discountType'] ?? 'percentage').toString(),
      discountValue: asNum(json['discountValue'])?.toDouble() ?? 0,
      minOrderValue: asNum(json['minOrderValue'])?.toDouble() ?? 0,
      maxDiscount: asNum(json['maxDiscount'])?.toDouble(),
      usageLimit: asNum(json['usageLimit'])?.toInt(),
      perUserLimit: asNum(json['perUserLimit'])?.toInt(),
      isFirstOrderOnly: json['isFirstOrderOnly'] == true,
      startDate: parseDate(json['startDate']),
      endDate: parseDate(json['endDate']),
      status: (json['status'] ?? 'active').toString(),
    );
  }

  final String id;
  final String couponCode;
  final String discountType; // percentage | flat-price
  final double discountValue;
  final double minOrderValue;
  final double? maxDiscount;
  final int? usageLimit;
  final int? perUserLimit;
  final bool isFirstOrderOnly;
  final DateTime? startDate;
  final DateTime? endDate;
  final String status;

  bool get isPercentage => discountType == 'percentage';
  bool get isActive => status == 'active';
}
